File: solutions/Java/binary_tree_zigzag_level_order_traversal_deque.java
import java.util.*;

public class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();
        if (root == null) return result;
        Queue<TreeNode> q = new LinkedList<>(); q.offer(root);
        int level = 0;
        while (!q.isEmpty()) {
            int sz = q.size(); Deque<Integer> dq = new LinkedList<>();
            for (int i = 0; i < sz; i++) {
                TreeNode n = q.poll();
                if (level % 2 == 0) dq.offer(n.val);
                else dq.offerFirst(n.val);
                if (n.left != null) q.offer(n.left);
                if (n.right != null) q.offer(n.right);
            }
            result.add(new ArrayList<>(dq));
            level++;
        }
        return result;
    }
    public static class TreeNode { int val; TreeNode left, right; TreeNode(int x) { val = x; } }
}

File: solutions/Rust/binary_tree_zigzag_level_order_traversal_deque.rs
use std::cell::RefCell; use std::collections::VecDeque; use std::rc::Rc;
#[derive(Debug)]
pub struct TreeNode { pub val: i32, pub left: Option<Rc<RefCell<TreeNode>>>, pub right: Option<Rc<RefCell<TreeNode>>>, }
pub fn zigzag_level_order(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<Vec<i32>> {
    let mut result = Vec::new(); if root.is_none() { return result; }
    let mut q = VecDeque::new(); q.push_back(root.unwrap());
    let mut level = 0;
    while !q.is_empty() { let sz = q.len(); let mut level_vals = VecDeque::new();
        for _ in 0..sz { let n = q.pop_front().unwrap(); let n_ref = n.borrow_mut();
            if level % 2 == 0 { level_vals.push_back(n_ref.val); }
            else { level_vals.push_front(n_ref.val); }
            if let Some(l) = n_ref.left.clone() { q.push_back(l); }
            if let Some(r) = n_ref.right.clone() { q.push_back(r); }
        }
        result.push(level_vals.into_iter().collect());
        level += 1;
    }
    result
}
